jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/WriteFlusher.java

```diff
@@ -512,6 +549,46 @@ public boolean onFail(Throwable cause)
         }
     }
 
+    public Callback cancel()
+    {
+        // Keep trying to handle the failure until we get to IDLE or FAILED state
+        while (true)
+        {
+            State current = _state.get();
+            switch (current.getType())
+            {
+                case IDLE:
+                case CANCELLED:
+                case FAILED:
+                    return null;
+
+                case PENDING:
+                    PendingState pending = (PendingState)current;
+                    if (updateState(current, __CANCELLED))
+                        return pending._callback;
+                    break;
+
+                case COMPLETING:
+                case WRITING:
+                    updateState(current, __CANCEL);
+                    break;
+
+                case CANCEL:
+                    Thread.onSpinWait();
+                    break;
+
+                case CANCELLING:
+                    CancellingState cancelling = (CancellingState)current;
+                    if (updateState(current, __CANCELLED))
+                        return cancelling._callback;
+                    break;
+
+                default:
+                    throw new IllegalStateException();
+            }
+        }
+    }
+
     public void onClose()
```

Comment:
I think this method needs to be updated with the cancel states, and so does `toStateString()`.

Reply:
`onFail` already handles the cancel states.